numpy/_core/src/common/npy_hashtable.c

```diff
@@ -228,7 +231,7 @@ PyArrayIdentityHash_SetItem(PyArrayIdentityHash *tb,
 
     PyObject **tb_item = find_item(tb, key);
     if (value != NULL) {
-        if (tb_item[0] != NULL && !replace) {
+        if (tb_item[0] != NULL && tb_item[0] != value && !replace) {
```

Comment:
It looks like all that's needed in the end is this identity check. Since the errors we were getting were for built-in dtypes I think the worst that can happen is a reference leak for an object that's already effectively immortal.